### ISSUE TITLE
GTP interface unresponsive at engine startup

### ISSUE THREAD

**Issue by evdwerf** [gcp/leela-zero#2152]
When starting Leela Zero it can take a long time for the engine to become responsive to GTP commands. This may be perceived negatively by users and can cause timeout issues in some GUI's (e.g., GridMaster on Android). The reason is OpenCL tuning and network loading, which is currently done immediately at startup (before the GTP loop starts) even though it is often not directly needed.

A relatively simple solution would be to delay the heavy lifting until it is really needed (or nothing else needs to be done). E.g., loading the network could be delayed until the first 'genmove' command is issued (as previously suggested here: https://github.com/gcp/leela-zero/issues/1751#issuecomment-432220261). Alternatively one could already start loading on the 'boardsize' command (as suggested here: https://github.com/gcp/leela-zero/issues/1679#issuecomment-454387448), though I'm not 100% sure if 'boardsize' is considered a mandatory command.

A more generic solution could be to define a policy to simply not allow any heavy lifting on the main/gui thread. Things like OpenCL tuning and network loading could then start immediately on a dedicated thread, with the main thread sleeping most of the time, and blocking only when a result from another thread is needed.

A combination of the above might be best. E.g. do the heavy lifting on a dedicated thread, but don't start until it's clear that the work is needed or the engine has nothing else left to do.

**Comment by gcp** [comment]
>maybe you could delete #define USE_OPENCL_SELFCHECK in config.h ?

I have no idea why you think this is at all relevant.

**Comment by gcp** [comment]
>E.g., loading the network could be delayed until the first 'genmove' command is issued

This doesn't "work" in the sense that if the game has a fast timecontrol, you might forfeit before you get to make a move. Regardless of the timecontrol you'd lose time at the start of the game.

>though I'm not 100% sure if 'boardsize' is considered a mandatory command.

The GTP spec says:
"There is no default state for any state variable. When first started, the engine may set these as it likes."

Now, if you don't send a boardsize, you have no idea what size board the engine is playing on. I can't imagine much meaningful interaction that requires a network to be loaded if the board size is not agreed, so I think this is *effectively* a required command before being able to do things like genmove etc.

**Comment by evdwerf** [comment]
Right, fast time controls are indeed a concern. For Leela Zero, as long as the board size is compiled-in, the network is not strictly needed to respond to the 'boardsize' command, but it might still make sense to block early somewhere around that point if the network loading is still busy loading.

If timing is the main concern we could also tie it to the 'time_settings' command (as long as we don't have time settings there is no need to worry about the clock), or perhaps just define a list of commands that will block until network loading is done regardless of which is called first.

**Comment by l1t1** [comment]
i misunstood evdwerfs question. i thought it was opencl checking caused the engine no response .just like https://github.com/gcp/leela-zero/issues/2136 said, because gridmaster used a modified wright

**Comment by gcp** [comment]
>If timing is the main concern we could also tie it to the 'time_settings' command (as long as we don't have time settings there is no need to worry about the clock)

Clever!
